### PR TITLE
Fix | Use cert chain to connect to IoT Core

### PR DESCRIPTION
**Pull request by frank10gm**
This fix enables the connection for android devices using a cert chain (without this JITP can't work)

Instead of reading just the first part of a certificate, we read the entire cert chain and use it to connect to IoT Core.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotKeystoreHelper.java [hunk 1]
```diff
@@ -291,13 +295,11 @@ private static KeyStore getTempKeystore(KeyStore customerKeystore, String certId
             KeyStore tempKeystore = KeyStore.getInstance(KeyStore.getDefaultType());
             tempKeystore.load(null);
 
-            X509Certificate cert = (X509Certificate) customerKeystore.getCertificate(certId);
-            tempKeystore.setCertificateEntry("cert-alias", cert);
+            X509Certificate[] certs = (X509Certificate[]) customerKeystore.getCertificateChain(certId);
```
Comment (raphkim) [inline aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotKeystoreHelper.java:298]: This is an illegal cast. `Keystore#getCertificateChain(String)` returns `Certificate[]` object, which is an array of superclass of `X509Certificate`, so casting the entire array will cause `ClassCastException` to be thrown. You would likely need to iterate through the list to cast individual item or use `Arrays.copyOf<T, U>(T[], int, Class<? extends U[]>)`.

**Review by raphkim — COMMENTED** [review]
Thank you @frank10gm for your contribution to AWS Mobile SDK! This looks like some promising stuff and it looks ready to go besides some minor comment to address.

However, I would really love to see some test suites accompany this PR before I can feel confident about merging it into the main branch. Please check out our guide for generating test resources [here](https://github.com/aws-amplify/amplify-ci-support/blob/main/src/integ_test_resources/android/sdk/integration/cdk/README.md) for assistance in writing/running integration tests locally on your device.

File: aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotKeystoreHelper.java (hunk 1 above)
Comment (frank10gm) [reply aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotKeystoreHelper.java]: Hi @raphkim, I fixed the illegal cast.

`Certificate[] certs = customerKeystore.getCertificateChain(certId);
            X509Certificate[] xcerts = new X509Certificate[certs.length];

            for (int i = 0; i < certs.length; i++) {
                xcerts[i] = (X509Certificate) certs[i];
            }`

**Comment by frank10gm** [comment]
Hi @raphkim 
in the meantime I executed the test suite, but I don't know in which format to post the result, sorry for my incompetence :)

I attach here the result as a screenshot: 
Screenshot 2021-06-29 at 10 54 11

**Comment by sdhuka** [comment]
> Hi @raphkim
> in the meantime I started cdk, but I don't know how to execute and in which format to post the result, sorry for my incompetence :)
> 
> I just ran cdk deploy iot, but I don't know how to move further
> Screenshot 2021-06-29 at 10 54 11
> 
> Can you please address me?
> 
> Just for information, this is the test result from Android Studio:
> [Test Results - All_in_aws-sdk-android_aws-android-sdk-iot.html.zip](https://github.com/aws-amplify/aws-sdk-android/files/6732392/Test.Results.-.All_in_aws-sdk-android_aws-android-sdk-iot.html.zip)

After deploying AWS test resources, please use the following commands in aws-sdk directory to generate test configuration. After the test configuration is generated run the integration tests.
export AWS_ACCESS_KEY_ID=YOUR_AWS_ACCESS_ID
export AWS_SECRET_ACCESS_KEY=YOUR_AWS_SECRET_ACCESS_KEY
export AWS_SESSION_TOKEN=YOUR_AWS_SESSION_TOKEN
cd aws-sdk-android
bash build-support/generate-test-config.sh -p android -r us-east-1
mkdir -p aws-android-sdk-testutils/src/main/res/raw
mkdir -p aws-android-sdk-mobile-client/src/androidTest/res/raw
ln -s $HOME/.aws-amplify/aws-sdk-android/testconfiguration.json aws-android-sdk-testutils/src/main/res/raw/testconfiguration.json
ln -s $HOME/.aws-amplify/aws-sdk-android/testconfiguration.json aws-android-sdk-mobile-client/src/androidTest/res/raw/awsconfiguration.json

**Comment by frank10gm** [comment]
Hi @sdhuka @raphkim ,
thank you for your precious help!

Here attached the integration tests results:

Screenshot 2021-07-01 at 19 34 38

[Test Results - MqttManagerIntegrati___.html.zip](https://github.com/aws-amplify/aws-sdk-android/files/6749878/Test.Results.-.MqttManagerIntegrati___.html.zip)

**Comment by raphkim** [comment]
Hi @frank10gm , thank you for the confirmation! Although it is certainly reassuring to see that the tests are still passing after the code changes, what we really look for is additional code coverage that specifically tests your use-case (JIT Provisioning). That said, I see the potential challenges in providing proper instrumentation tests with limited resources available. My suggestion at this point would be to build the SDK with your code changes, publish it to local maven, and then test it in your app to manually confirm that this PR address your needs. 

Please understand that while we are eager to accept user contributions, we must still insist on maintaining the highest quality in our projects. Please refer to our [contributing docs](https://github.com/aws-amplify/aws-sdk-android/blob/8de48797ae2e5c157169a94746c768da7de61056/.github/CONTRIBUTING.md) for more information. (Here is our [more detailed, up-to-date document](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md) from the `amplify-android` repo).

Feel free to reach out to us for any help with this process and we will be happy to assist.

**Comment by frank10gm** [comment]
> Hi @frank10gm , thank you for the confirmation! Although it is certainly reassuring to see that the tests are still passing after the code changes, what we really look for is additional code coverage that specifically tests your use-case (JIT Provisioning). That said, I see the potential challenges in providing proper instrumentation tests with limited resources available. My suggestion at this point would be to build the SDK with your code changes, publish it to local maven, and then test it in your app to manually confirm that this PR addresses your needs.
> 
> Please understand that while we are eager to accept user contributions, we must still insist on maintaining the highest quality in our projects. Please refer to our [contributing docs](https://github.com/aws-amplify/aws-sdk-android/blob/8de48797ae2e5c157169a94746c768da7de61056/.github/CONTRIBUTING.md) for more information. (Here is our [more detailed, up-to-date document](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md) from the `amplify-android` repo).
> 
> Feel free to reach out to us for any help with this process and we will be happy to assist.

Hi @raphkim ,
I fully understand your point and totally agree about the need of test coverage for JITP! I think I can't manage to cover that part by myself right now.

Anyway I already built the sdk with my change and tested it successfully, furthermore my team is using my forked repo for a production ready appliance that's close to go to the market. 
My fork was also tested with success by our client (a big international IoT company, I can put you in contact with them). We are also in contact with European IoT AWS team (Angelino Massimiliano) and addressed our code change also to them.

I can show you that my code is working good with JITP, just tell me what kind of demonstration I can provide. I will be happy to send you anything you need and, when I can, I will also be happy to provide test coverage of this part.

Have a nice day

**Review by raphkim — APPROVED** [review]

**Pull request merged**